File: go/noise/conn.go
package noise

import (
	"errors"
	"sync"
	"time"
)

// ConnState represents the state of a connection.
type ConnState int

const (
	// ConnStateNew indicates a newly created connection.
	ConnStateNew ConnState = iota
	// ConnStateHandshaking indicates the connection is performing handshake.
	ConnStateHandshaking
	// ConnStateEstablished indicates the connection is ready for data transfer.
	ConnStateEstablished
	// ConnStateClosed indicates the connection has been closed.
	ConnStateClosed
)

func (s ConnState) String() string {
	switch s {
	case ConnStateNew:
		return "new"
	case ConnStateHandshaking:
		return "handshaking"
	case ConnStateEstablished:
		return "established"
	case ConnStateClosed:
		return "closed"
	default:
		return "unknown"
	}
}

// Conn represents a connection to a remote peer.
// It manages the handshake process and provides a simple API
// for sending and receiving encrypted messages.
type Conn struct {
	mu sync.RWMutex

	// Configuration
	localKey   *KeyPair
	remotePK   PublicKey
	transport  Transport
	remoteAddr Addr

	// State
	state    ConnState
	session  *Session
	hsState  *HandshakeState
	localIdx uint32

	// Timestamps
	createdAt time.Time

	// Inbound channel for listener-managed connections
	// When set, Recv() reads from this channel instead of the transport
	inbound chan inboundPacket
}

// inboundPacket represents a packet received from the listener
type inboundPacket struct {
	data []byte
	addr Addr
}

// ConnConfig contains the configuration for creating a connection.
type ConnConfig struct {
	// LocalKey is the local static key pair.
	LocalKey *KeyPair
	// RemotePK is the remote peer's public key (required for initiator).
	RemotePK PublicKey
	// Transport is the underlying datagram transport.
	Transport Transport
	// RemoteAddr is the remote peer's address.
	RemoteAddr Addr
}

// NewConn creates a new connection with the given configuration.
// The connection is not yet established; call Dial() or process incoming
// handshake messages to complete the handshake.
func NewConn(cfg ConnConfig) (*Conn, error) {
	if cfg.LocalKey == nil {
		return nil, ErrMissingLocalKey
	}
	if cfg.Transport == nil {
		return nil, ErrMissingTransport
	}

	localIdx, err := GenerateIndex()
	if err != nil {
		return nil, err
	}

	return &Conn{
		localKey:   cfg.LocalKey,
		remotePK:   cfg.RemotePK,
		transport:  cfg.Transport,
		remoteAddr: cfg.RemoteAddr,
		state:      ConnStateNew,
		localIdx:   localIdx,
		createdAt:  time.Now(),
	}, nil
}

// Open initiates a handshake with the remote peer.
// This is a blocking call that completes the full handshake.
// The connection must have RemotePK and RemoteAddr configured.
func (c *Conn) Open() error {
	c.mu.Lock()
	if c.state != ConnStateNew {
		c.mu.Unlock()
		return ErrInvalidConnState
	}
	if c.remotePK.IsZero() {
		c.mu.Unlock()
		return ErrMissingRemotePK
	}
	if c.remoteAddr == nil {
		c.mu.Unlock()
		return ErrMissingRemoteAddr
	}

	c.state = ConnStateHandshaking

	// Create handshake state (IK pattern - we know remote's public key)
	hs, err := NewHandshakeState(Config{
		Pattern:      PatternIK,
		Initiator:    true,
		LocalStatic:  c.localKey,
		RemoteStatic: &c.remotePK,
	})
	if err != nil {
		c.state = ConnStateNew
		c.mu.Unlock()
		return err
	}
	c.hsState = hs
	c.mu.Unlock()

	// Generate and send handshake initiation
	msg1, err := hs.WriteMessage(nil)
	if err != nil {
		return c.failHandshake(err)
	}

	// msg1 format: ephemeral(32) + encrypted_static(48) = 80 bytes
	// Build wire message
	wireMsg := BuildHandshakeInit(c.localIdx, hs.LocalEphemeral(), msg1[KeySize:])
	if err := c.transport.SendTo(wireMsg, c.remoteAddr); err != nil {
		return c.failHandshake(err)
	}

	// Wait for handshake response
	buf := make([]byte, MaxPacketSize)
	n, _, err := c.transport.RecvFrom(buf)
	if err != nil {
		return c.failHandshake(err)
	}

	// Parse response
	resp, err := ParseHandshakeResp(buf[:n])
	if err != nil {
		return c.failHandshake(err)
	}

	// Verify receiver index matches our sender index
	if resp.ReceiverIndex != c.localIdx {
		return c.failHandshake(ErrInvalidReceiverIndex)
	}

	// Reconstruct the noise message and process
	noiseMsg := make([]byte, KeySize+16)
	copy(noiseMsg[:KeySize], resp.Ephemeral[:])
	copy(noiseMsg[KeySize:], resp.Empty)

	if _, err := hs.ReadMessage(noiseMsg); err != nil {
		return c.failHandshake(err)
	}

	// Complete handshake
	return c.completeHandshake(resp.SenderIndex, nil)
}

// Accept processes an incoming handshake initiation and completes the handshake.
// This is used by the responder to accept incoming connections.
// Returns the handshake response to send back.
func (c *Conn) Accept(msg *HandshakeInitMessage) ([]byte, error) {
	c.mu.Lock()
	if c.state != ConnStateNew {
		c.mu.Unlock()
		return nil, ErrInvalidConnState
	}

	c.state = ConnStateHandshaking

	// Create handshake state (IK pattern - responder)
	hs, err := NewHandshakeState(Config{
		Pattern:     PatternIK,
		Initiator:   false,
		LocalStatic: c.localKey,
	})
	if err != nil {
		c.state = ConnStateNew
		c.mu.Unlock()
		return nil, err
	}
	c.hsState = hs
	c.mu.Unlock()

	// Reconstruct the noise message: ephemeral(32) + static_enc(48) = 80 bytes
	noiseMsg := make([]byte, KeySize+48)
	copy(noiseMsg[:KeySize], msg.Ephemeral[:])
	copy(noiseMsg[KeySize:], msg.Static)

	if _, err := hs.ReadMessage(noiseMsg); err != nil {
		return nil, c.failHandshake(err)
	}

	// Get remote public key from handshake
	remotePK := hs.RemoteStatic()

	// Generate response
	msg2, err := hs.WriteMessage(nil)
	if err != nil {
		return nil, c.failHandshake(err)
	}

	// Store initiator's index as remote index
	remoteIdx := msg.SenderIndex

	// Complete handshake (updates remotePK atomically with state)
	if err := c.completeHandshake(remoteIdx, &remotePK); err != nil {
		return nil, err
	}

	// Build wire response message
	return BuildHandshakeResp(c.localIdx, remoteIdx, hs.LocalEphemeral(), msg2[KeySize:]), nil
}

// completeHandshake finalizes the handshake and creates the session.
// If remotePK is not nil, it will be set atomically with the state transition.
func (c *Conn) completeHandshake(remoteIdx uint32, remotePK *PublicKey) error {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.hsState == nil || !c.hsState.IsFinished() {
		return ErrHandshakeIncomplete
	}

	// Get transport keys
	sendCS, recvCS, err := c.hsState.Split()
	if err != nil {
		c.state = ConnStateNew
		return err
	}

	// Update remotePK if provided (for responder case)
	if remotePK != nil {
		c.remotePK = *remotePK
	}

	// Create session
	session, err := NewSession(SessionConfig{
		LocalIndex:  c.localIdx,
		RemoteIndex: remoteIdx,
		SendKey:     sendCS.Key(),
		RecvKey:     recvCS.Key(),
		RemotePK:    c.remotePK,
	})
	if err != nil {
		c.state = ConnStateNew
		return err
	}

	c.session = session
	c.hsState = nil // Clear handshake state
	c.state = ConnStateEstablished

	return nil
}

// failHandshake handles handshake failure.
func (c *Conn) failHandshake(err error) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.state = ConnStateNew
	c.hsState = nil
	return err
}

// Send sends an encrypted message to the remote peer.
// The protocol byte indicates the type of payload.
func (c *Conn) Send(protocol byte, payload []byte) error {
	c.mu.RLock()
	if c.state != ConnStateEstablished {
		c.mu.RUnlock()
		return ErrNotEstablished
	}
	session := c.session
	remoteAddr := c.remoteAddr
	c.mu.RUnlock()

	// Encode payload with protocol byte
	plaintext := EncodePayload(protocol, payload)

	// Encrypt
	ciphertext, counter, err := session.Encrypt(plaintext)
	if err != nil {
		return err
	}

	// Build wire message
	msg := BuildTransportMessage(session.RemoteIndex(), counter, ciphertext)

	// Send
	return c.transport.SendTo(msg, remoteAddr)
}

// Recv receives and decrypts a message from the remote peer.
// Returns the protocol byte and decrypted payload.
// This is a blocking call.
func (c *Conn) Recv() (protocol byte, payload []byte, err error) {
	c.mu.RLock()
	if c.state != ConnStateEstablished {
		c.mu.RUnlock()
		return 0, nil, ErrNotEstablished
	}
	session := c.session
	inbound := c.inbound
	c.mu.RUnlock()

	var data []byte

	if inbound != nil {
		// Listener-managed connection: read from inbound channel
		pkt, ok := <-inbound
		if !ok {
			return 0, nil, ErrConnClosed
		}
		data = pkt.data
	} else {
		// Direct connection: read from transport
		buf := make([]byte, MaxPacketSize)
		n, _, err := c.transport.RecvFrom(buf)
		if err != nil {
			return 0, nil, err
		}
		data = buf[:n]
	}

	// Parse transport message
	msg, err := ParseTransportMessage(data)
	if err != nil {
		return 0, nil, err
	}

	// Verify receiver index
	if msg.ReceiverIndex != session.LocalIndex() {
		return 0, nil, ErrInvalidReceiverIndex
	}

	// Decrypt
	plaintext, err := session.Decrypt(msg.Ciphertext, msg.Counter)
	if err != nil {
		return 0, nil, err
	}

	// Decode protocol and payload
	return DecodePayload(plaintext)
}

// Close closes the connection.
func (c *Conn) Close() error {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.state == ConnStateClosed {
		return nil
	}

	c.state = ConnStateClosed
	if c.session != nil {
		c.session.Expire()
	}

	return nil
}

// State returns the current connection state.
func (c *Conn) State() ConnState {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.state
}

// RemotePublicKey returns the remote peer's public key.
func (c *Conn) RemotePublicKey() PublicKey {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.remotePK
}

// RemoteAddr returns the remote peer's address.
func (c *Conn) RemoteAddr() Addr {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.remoteAddr
}

// LocalIndex returns the local session index.
func (c *Conn) LocalIndex() uint32 {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.localIdx
}

// Session returns the underlying session (nil if not established).
func (c *Conn) Session() *Session {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.session
}

// SetRemoteAddr updates the remote address (for NAT traversal).
func (c *Conn) SetRemoteAddr(addr Addr) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.remoteAddr = addr
}

// setInbound sets up the inbound channel for listener-managed connections.
// This should only be called by Listener before the connection is returned.
func (c *Conn) setInbound(ch chan inboundPacket) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.inbound = ch
}

// deliverPacket delivers a packet to the connection's inbound channel.
// Returns false if the channel is full or the connection is closed.
func (c *Conn) deliverPacket(data []byte, addr Addr) bool {
	c.mu.RLock()
	inbound := c.inbound
	state := c.state
	c.mu.RUnlock()

	if inbound == nil || state == ConnStateClosed {
		return false
	}

	select {
	case inbound <- inboundPacket{data: data, addr: addr}:
		return true
	default:
		return false
	}
}

// Connection errors.
var (
	ErrMissingLocalKey      = errors.New("noise: missing local key pair")
	ErrMissingTransport     = errors.New("noise: missing transport")
	ErrMissingRemotePK      = errors.New("noise: missing remote public key")
	ErrMissingRemoteAddr    = errors.New("noise: missing remote address")
	ErrInvalidConnState     = errors.New("noise: invalid connection state")
	ErrNotEstablished       = errors.New("noise: connection not established")
	ErrInvalidReceiverIndex = errors.New("noise: invalid receiver index")
	ErrHandshakeIncomplete  = errors.New("noise: handshake not complete")
	ErrConnClosed           = errors.New("noise: connection closed")
)

File: rust/src/conn.rs
//! Connection management for Noise-based communication.
//!
//! This module provides the `Conn` type which manages the handshake process
//! and provides a simple API for sending and receiving encrypted messages.

use std::sync::RwLock;

use crate::handshake::{Config, HandshakeState, Pattern};
use crate::keypair::{Key, KeyPair};
use crate::message::{
    build_handshake_init, build_handshake_resp, build_transport_message, decode_payload,
    encode_payload, parse_handshake_resp, parse_transport_message, HandshakeInit,
    KEY_SIZE, MAX_PACKET_SIZE,
};
use crate::session::{generate_index, Session, SessionConfig};
use crate::transport::{Addr, Transport, TransportError};

/// Connection state.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ConnState {
    /// Newly created connection.
    New,
    /// Handshake in progress.
    Handshaking,
    /// Connection established, ready for data transfer.
    Established,
    /// Connection closed.
    Closed,
}

impl std::fmt::Display for ConnState {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::New => write!(f, "new"),
            Self::Handshaking => write!(f, "handshaking"),
            Self::Established => write!(f, "established"),
            Self::Closed => write!(f, "closed"),
        }
    }
}

/// Connection errors.
#[derive(Debug)]
pub enum ConnError {
    /// Missing local key pair.
    MissingLocalKey,
    /// Missing transport.
    MissingTransport,
    /// Missing remote public key.
    MissingRemotePK,
    /// Missing remote address.
    MissingRemoteAddr,
    /// Invalid connection state.
    InvalidState,
    /// Connection not established.
    NotEstablished,
    /// Invalid receiver index.
    InvalidReceiverIndex,
    /// Handshake not complete.
    HandshakeIncomplete,
    /// Handshake error.
    Handshake(crate::handshake::Error),
    /// Session error.
    Session(crate::session::SessionError),
    /// Message error.
    Message(crate::message::MessageError),
    /// Transport error.
    Transport(TransportError),
}

impl std::fmt::Display for ConnError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::MissingLocalKey => write!(f, "missing local key pair"),
            Self::MissingTransport => write!(f, "missing transport"),
            Self::MissingRemotePK => write!(f, "missing remote public key"),
            Self::MissingRemoteAddr => write!(f, "missing remote address"),
            Self::InvalidState => write!(f, "invalid connection state"),
            Self::NotEstablished => write!(f, "connection not established"),
            Self::InvalidReceiverIndex => write!(f, "invalid receiver index"),
            Self::HandshakeIncomplete => write!(f, "handshake not complete"),
            Self::Handshake(e) => write!(f, "handshake error: {}", e),
            Self::Session(e) => write!(f, "session error: {}", e),
            Self::Message(e) => write!(f, "message error: {}", e),
            Self::Transport(e) => write!(f, "transport error: {}", e),
        }
    }
}

impl std::error::Error for ConnError {}

impl From<crate::handshake::Error> for ConnError {
    fn from(e: crate::handshake::Error) -> Self {
        Self::Handshake(e)
    }
}

impl From<crate::session::SessionError> for ConnError {
    fn from(e: crate::session::SessionError) -> Self {
        Self::Session(e)
    }
}

impl From<crate::message::MessageError> for ConnError {
    fn from(e: crate::message::MessageError) -> Self {
        Self::Message(e)
    }
}

impl From<TransportError> for ConnError {
    fn from(e: TransportError) -> Self {
        Self::Transport(e)
    }
}

/// Result type for connection operations.
pub type Result<T> = std::result::Result<T, ConnError>;

/// Configuration for creating a connection.
pub struct ConnConfig<T: Transport + 'static> {
    /// Local static key pair.
    pub local_key: KeyPair,
    /// Remote peer's public key (required for initiator).
    pub remote_pk: Option<Key>,
    /// Underlying datagram transport.
    pub transport: T,
    /// Remote peer's address.
    pub remote_addr: Option<Box<dyn Addr>>,
}

/// A connection to a remote peer.
///
/// Manages the handshake process and provides a simple API
/// for sending and receiving encrypted messages.
pub struct Conn<T: Transport + 'static> {
    // Configuration
    local_key: KeyPair,
    remote_pk: RwLock<Key>,
    transport: T,
    remote_addr: RwLock<Option<Box<dyn Addr>>>,

    // State
    state: RwLock<ConnState>,
    session: RwLock<Option<Session>>,
    local_idx: u32,
}

impl<T: Transport + 'static> Conn<T> {
    /// Creates a new connection with the given configuration.
    pub fn new(cfg: ConnConfig<T>) -> Result<Self> {
        let local_idx = generate_index();

        Ok(Self {
            local_key: cfg.local_key,
            remote_pk: RwLock::new(cfg.remote_pk.unwrap_or_default()),
            transport: cfg.transport,
            remote_addr: RwLock::new(cfg.remote_addr),
            state: RwLock::new(ConnState::New),
            session: RwLock::new(None),
            local_idx,
        })
    }

    /// Initiates a handshake with the remote peer.
    /// This is a blocking call that completes the full handshake.
    pub fn open(&self) -> Result<()> {
        // Read required values first (read locks are cheap)
        let remote_pk = *self.remote_pk.read().unwrap();
        if remote_pk.is_zero() {
            return Err(ConnError::MissingRemotePK);
        }
        {
            let remote_addr = self.remote_addr.read().unwrap();
            if remote_addr.is_none() {
                return Err(ConnError::MissingRemoteAddr);
            }
        }

        // Now atomically check and update state (minimizing write lock time)
        {
            let mut state = self.state.write().unwrap();
            if *state != ConnState::New {
                return Err(ConnError::InvalidState);
            }
            *state = ConnState::Handshaking;
        }

        // Create handshake state (IK pattern)
        let hs_result = HandshakeState::new(Config {
            pattern: Some(Pattern::IK),
            initiator: true,
            local_static: Some(self.local_key.clone()),
            remote_static: Some(remote_pk),
            ..Default::default()
        });
        let mut hs = match hs_result {
            Ok(hs) => hs,
            Err(e) => {
                self.fail_handshake();
                return Err(e.into());
            }
        };

        // Generate and send handshake initiation
        let msg1 = match hs.write_message(&[]) {
            Ok(msg) => msg,
            Err(e) => {
                self.fail_handshake();
                return Err(e.into());
            }
        };

        // msg1 format: ephemeral(32) + encrypted_static(48) = 80 bytes
        let wire_msg = build_handshake_init(
            self.local_idx,
            &hs.local_ephemeral().unwrap(),
            &msg1[KEY_SIZE..],
        );

        let remote_addr = self.remote_addr.read().unwrap();
        if let Err(e) = self
            .transport
            .send_to(&wire_msg, remote_addr.as_ref().unwrap().as_ref())
        {
            self.fail_handshake();
            return Err(e.into());
        }

        // Wait for handshake response
        let mut buf = vec![0u8; MAX_PACKET_SIZE];
        let (n, _) = match self.transport.recv_from(&mut buf) {
            Ok(result) => result,
            Err(e) => {
                self.fail_handshake();
                return Err(e.into());
            }
        };

        // Parse response
        let resp = match parse_handshake_resp(&buf[..n]) {
            Ok(resp) => resp,
            Err(e) => {
                self.fail_handshake();
                return Err(e.into());
            }
        };

        // Verify receiver index matches our sender index
        if resp.receiver_index != self.local_idx {
            self.fail_handshake();
            return Err(ConnError::InvalidReceiverIndex);
        }

        // Reconstruct the noise message and process
        let mut noise_msg = vec![0u8; KEY_SIZE + 16];
        noise_msg[..KEY_SIZE].copy_from_slice(resp.ephemeral.as_bytes());
        noise_msg[KEY_SIZE..].copy_from_slice(&resp.empty_encrypted);

        if let Err(e) = hs.read_message(&noise_msg) {
            self.fail_handshake();
            return Err(e.into());
        }

        // Complete handshake
        self.complete_handshake(&hs, resp.sender_index, None)
    }

    /// Processes an incoming handshake initiation and completes the handshake.
    /// Returns the handshake response to send back.
    pub fn accept(&self, msg: &HandshakeInit) -> Result<Vec<u8>> {
        // Check and update state
        {
            let mut state = self.state.write().unwrap();
            if *state != ConnState::New {
                return Err(ConnError::InvalidState);
            }
            *state = ConnState::Handshaking;
        }

        // Create handshake state (IK pattern - responder)
        let mut hs = match HandshakeState::new(Config {
            pattern: Some(Pattern::IK),
            initiator: false,
            local_static: Some(self.local_key.clone()),
            ..Default::default()
        }) {
            Ok(hs) => hs,
            Err(e) => {
                self.fail_handshake();
                return Err(e.into());
            }
        };

        // Reconstruct the noise message: ephemeral(32) + static_enc(48) = 80 bytes
        let mut noise_msg = vec![0u8; KEY_SIZE + 48];
        noise_msg[..KEY_SIZE].copy_from_slice(msg.ephemeral.as_bytes());
        noise_msg[KEY_SIZE..].copy_from_slice(&msg.static_encrypted);

        if let Err(e) = hs.read_message(&noise_msg) {
            self.fail_handshake();
            return Err(e.into());
        }

        // Get remote public key from handshake
        let remote_pk = *hs.remote_static();

        // Generate response
        let msg2 = match hs.write_message(&[]) {
            Ok(msg) => msg,
            Err(e) => {
                self.fail_handshake();
                return Err(e.into());
            }
        };

        // Store initiator's index as remote index
        let remote_idx = msg.sender_index;

        // Complete handshake (updates remote_pk atomically with state)
        self.complete_handshake(&hs, remote_idx, Some(remote_pk))?;

        // Build wire response message
        Ok(build_handshake_resp(
            self.local_idx,
            remote_idx,
            &hs.local_ephemeral().unwrap(),
            &msg2[KEY_SIZE..],
        ))
    }

    /// Completes the handshake and creates the session.
    /// If remote_pk is provided, it will be set atomically with the state transition.
    fn complete_handshake(&self, hs: &HandshakeState, remote_idx: u32, remote_pk: Option<Key>) -> Result<()> {
        if !hs.is_finished() {
            return Err(ConnError::HandshakeIncomplete);
        }

        // Get transport keys
        let (send_cipher, recv_cipher) = hs.split()?;

        // Update remote_pk if provided (for responder case)
        if let Some(pk) = remote_pk {
            let mut remote_pk_lock = self.remote_pk.write().unwrap();
            *remote_pk_lock = pk;
        }

        // Create session
        let current_remote_pk = *self.remote_pk.read().unwrap();
        let session = Session::new(SessionConfig {
            local_index: self.local_idx,
            remote_index: remote_idx,
            send_key: *send_cipher.key(),
            recv_key: *recv_cipher.key(),
            remote_pk: current_remote_pk,
        });

        // Update state
        {
            let mut session_lock = self.session.write().unwrap();
            *session_lock = Some(session);
        }
        {
            let mut state = self.state.write().unwrap();
            *state = ConnState::Established;
        }

        Ok(())
    }

    /// Handles handshake failure.
    fn fail_handshake(&self) {
        let mut state = self.state.write().unwrap();
        *state = ConnState::New;
    }

    /// Sends an encrypted message to the remote peer.
    pub fn send(&self, protocol: u8, payload: &[u8]) -> Result<()> {
        // Check state and get session info
        let (ciphertext, counter, remote_index) = {
            let state = self.state.read().unwrap();
            if *state != ConnState::Established {
                return Err(ConnError::NotEstablished);
            }

            let session_lock = self.session.read().unwrap();
            let session = session_lock.as_ref().unwrap();

            // Encode and encrypt
            let plaintext = encode_payload(protocol, payload);
            let (ciphertext, counter) = session.encrypt(&plaintext)?;
            (ciphertext, counter, session.remote_index())
        };

        // Get remote address
        let remote_addr = self.remote_addr.read().unwrap();
        let remote_addr = remote_addr.as_ref().ok_or(ConnError::MissingRemoteAddr)?;

        // Build and send message
        let msg = build_transport_message(remote_index, counter, &ciphertext);
        self.transport.send_to(&msg, remote_addr.as_ref())?;

        Ok(())
    }

    /// Receives and decrypts a message from the remote peer.
    /// Returns the protocol byte and decrypted payload.
    ///
    /// For better performance in tight loops, use `recv_with_buffer` with a reusable buffer.
    pub fn recv(&self) -> Result<(u8, Vec<u8>)> {
        let mut buf = vec![0u8; MAX_PACKET_SIZE];
        self.recv_with_buffer(&mut buf)
    }

    /// Receives and decrypts a message using the provided buffer.
    /// This avoids allocating a new buffer on each call.
    /// The buffer should be at least MAX_PACKET_SIZE bytes.
    pub fn recv_with_buffer(&self, buf: &mut [u8]) -> Result<(u8, Vec<u8>)> {
        {
            let state = self.state.read().unwrap();
            if *state != ConnState::Established {
                return Err(ConnError::NotEstablished);
            }
        }

        // Receive packet
        let (n, _) = self.transport.recv_from(buf)?;

        // Parse transport message
        let msg = parse_transport_message(&buf[..n])?;

        // Verify receiver index
        if msg.receiver_index != self.local_idx {
            return Err(ConnError::InvalidReceiverIndex);
        }

        // Decrypt
        let plaintext = {
            let session_lock = self.session.read().unwrap();
            let session = session_lock.as_ref().unwrap();
            session.decrypt(msg.ciphertext, msg.counter)?
        };

        // Decode protocol and payload
        let (protocol, payload) = decode_payload(&plaintext)?;
        Ok((protocol, payload.to_vec()))
    }

    /// Closes the connection.
    pub fn close(&self) -> Result<()> {
        let mut state = self.state.write().unwrap();
        if *state == ConnState::Closed {
            return Ok(());
        }

        *state = ConnState::Closed;
        if let Some(session) = self.session.write().unwrap().as_mut() {
            session.expire();
        }

        Ok(())
    }

    /// Returns the current connection state.
    pub fn state(&self) -> ConnState {
        *self.state.read().unwrap()
    }

    /// Returns the remote peer's public key.
    pub fn remote_public_key(&self) -> Key {
        *self.remote_pk.read().unwrap()
    }

    /// Returns the local session index.
    pub fn local_index(&self) -> u32 {
        self.local_idx
    }

    /// Updates the remote address (for NAT traversal).
    pub fn set_remote_addr(&self, addr: Box<dyn Addr>) {
        let mut remote_addr = self.remote_addr.write().unwrap();
        *remote_addr = Some(addr);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::message::parse_handshake_init;
    use crate::transport::MockTransport;
    use std::sync::Arc;

    #[test]
    fn test_conn_state_display() {
        assert_eq!(ConnState::New.to_string(), "new");
        assert_eq!(ConnState::Handshaking.to_string(), "handshaking");
        assert_eq!(ConnState::Established.to_string(), "established");
        assert_eq!(ConnState::Closed.to_string(), "closed");
    }

    #[test]
    fn test_conn_new() {
        let key = KeyPair::generate();
        let transport = MockTransport::new("test");
        
        let conn = Conn::new(ConnConfig {
            local_key: key,
            remote_pk: None,
            transport: Arc::clone(&transport),
            remote_addr: None,
        }).unwrap();

        assert_eq!(conn.state(), ConnState::New);
        assert!(conn.local_index() != 0);
    }

    #[test]
    fn test_conn_open_missing_remote_pk() {
        let key = KeyPair::generate();
        let transport = MockTransport::new("test");
        
        let conn = Conn::new(ConnConfig {
            local_key: key,
            remote_pk: None,
            transport: Arc::clone(&transport),
            remote_addr: Some(Box::new(crate::transport::MockAddr::new("peer"))),
        }).unwrap();

        let err = conn.open().unwrap_err();
        assert!(matches!(err, ConnError::MissingRemotePK));
    }

    #[test]
    fn test_conn_open_missing_remote_addr() {
        let key = KeyPair::generate();
        let peer_key = KeyPair::generate();
        let transport = MockTransport::new("test");
        
        let conn = Conn::new(ConnConfig {
            local_key: key,
            remote_pk: Some(peer_key.public),
            transport: Arc::clone(&transport),
            remote_addr: None,
        }).unwrap();

        let err = conn.open().unwrap_err();
        assert!(matches!(err, ConnError::MissingRemoteAddr));
    }

    #[test]
    fn test_conn_send_not_established() {
        let key = KeyPair::generate();
        let transport = MockTransport::new("test");
        
        let conn = Conn::new(ConnConfig {
            local_key: key,
            remote_pk: None,
            transport: Arc::clone(&transport),
            remote_addr: None,
        }).unwrap();

        let err = conn.send(0, b"test").unwrap_err();
        assert!(matches!(err, ConnError::NotEstablished));
    }

    #[test]
    fn test_conn_recv_not_established() {
        let key = KeyPair::generate();
        let transport = MockTransport::new("test");
        
        let conn = Conn::new(ConnConfig {
            local_key: key,
            remote_pk: None,
            transport: Arc::clone(&transport),
            remote_addr: None,
        }).unwrap();

        let err = conn.recv().unwrap_err();
        assert!(matches!(err, ConnError::NotEstablished));
    }

    #[test]
    fn test_conn_close() {
        let key = KeyPair::generate();
        let transport = MockTransport::new("test");
        
        let conn = Conn::new(ConnConfig {
            local_key: key,
            remote_pk: None,
            transport: Arc::clone(&transport),
            remote_addr: None,
        }).unwrap();

        conn.close().unwrap();
        assert_eq!(conn.state(), ConnState::Closed);

        // Double close should be ok
        conn.close().unwrap();
    }

    #[test]
    fn test_conn_handshake_and_communication() {
        let initiator_key = KeyPair::generate();
        let responder_key = KeyPair::generate();

        let initiator_transport = MockTransport::new("initiator");
        let responder_transport = MockTransport::new("responder");
        MockTransport::connect(&initiator_transport, &responder_transport);

        let initiator = Conn::new(ConnConfig {
            local_key: initiator_key,
            remote_pk: Some(responder_key.public),
            transport: Arc::clone(&initiator_transport),
            remote_addr: Some(Box::new(crate::transport::MockAddr::new("responder"))),
        }).unwrap();

        let responder = Conn::new(ConnConfig {
            local_key: responder_key,
            remote_pk: None,
            transport: Arc::clone(&responder_transport),
            remote_addr: Some(Box::new(crate::transport::MockAddr::new("initiator"))),
        }).unwrap();

        // Spawn responder thread
        let responder_handle = std::thread::spawn(move || {
            // Receive handshake init
            let mut buf = vec![0u8; MAX_PACKET_SIZE];
            let (n, _) = responder_transport.recv_from(&mut buf).unwrap();
            let init_msg = parse_handshake_init(&buf[..n]).unwrap();

            // Process and respond
            let resp = responder.accept(&init_msg).unwrap();
            responder_transport.send_to(&resp, &crate::transport::MockAddr::new("initiator")).unwrap();

            responder
        });

        // Give responder time to start
        std::thread::sleep(std::time::Duration::from_millis(10));

        // Initiator opens connection
        initiator.open().unwrap();

        let responder = responder_handle.join().unwrap();

        // Verify both sides are established
        assert_eq!(initiator.state(), ConnState::Established);
        assert_eq!(responder.state(), ConnState::Established);

        // Test communication: initiator -> responder
        initiator.send(crate::message::protocol::CHAT, b"Hello from initiator!").unwrap();
        let (proto, payload) = responder.recv().unwrap();
        assert_eq!(proto, crate::message::protocol::CHAT);
        assert_eq!(payload, b"Hello from initiator!");

        // Test communication: responder -> initiator
        responder.send(crate::message::protocol::RPC, b"Hello from responder!").unwrap();
        let (proto, payload) = initiator.recv().unwrap();
        assert_eq!(proto, crate::message::protocol::RPC);
        assert_eq!(payload, b"Hello from responder!");
    }
}
